src/plugins/intel_cpu/src/nodes/mkldnn_roi_align_node.cpp

```diff
@@ -299,68 +975,112 @@ void MKLDNNROIAlignNode::executeSpecified() {
-                                    pointVector[sampleIndex + 2].second * wInputStride + blockResidual_;
-                float part3 = srcData[part3Index];
-                size_t part4Index = binOffsetInput_ + pointVector[sampleIndex + 3].first * hInputStride +
-                                    pointVector[sampleIndex + 3].second * wInputStride + blockResidual_;
-                float part4 = srcData[part4Index];
-
-                float sampleValue =
-                        weightVector[sampleIndex] * part1 +
-                        weightVector[sampleIndex + 1] * part2 +
-                        weightVector[sampleIndex + 2] * part3 +
-                        weightVector[sampleIndex + 3] * part4;
-                switch (getAlgorithm()) {
-                    case Algorithm::ROIAlignMax:
-                    {
-                        pooledValue = sampleValue > pooledValue ? sampleValue : pooledValue;
-                        break;
-                    }
-                    case Algorithm::ROIAlignAvg:
-                    default:
-                    {
-                        pooledValue += sampleValue / numSamplesInBin;
+
+        const int interp_values = 4;
+        // batch offset
+        size_t batchSrcOffset = roiBatchInd * batchInputStride;
+        size_t batchDstOffset = n * batchOutputStride;
+        float numSamplesInBinInvert = 1.f / numSamplesInBin;
+
+        if (roi_align_kernel) {
+            if (!isPlainFmt) {
+                parallel_for2d(pooledH, pooledW, [&](int yBinInd, int xBinInd) {
+                    // each sample have 4 values for Y, X, weight
+                    size_t binOffset = numSamplesInBin * interp_values * pooledW * yBinInd + numSamplesInBin * interp_values * xBinInd;
+
+                    std::vector<size_t> srcAddressList(numSamplesInBin * interp_values, 0);
+                    for (int sample = 0; sample < numSamplesInBin; sample++) {
+                        for (int i = 0; i < interp_values; i++) {
+                            size_t srcOffset = batchSrcOffset +
+                                pointVectorY[binOffset + sample * interp_values + i] * W * lastBlockDim +
+                                pointVectorX[binOffset + sample * interp_values + i] * lastBlockDim;
+                            srcAddressList[sample * interp_values + i] = reinterpret_cast<size_t>(&srcData[srcOffset]);
+                        }
                     }
+
+                    auto arg = jit_roi_align_call_args();
+                    arg.src = static_cast<const void*>(&srcAddressList[0]);
+                    arg.weights = static_cast<const float*>(&weightVector[binOffset]);
+
+                    arg.work_amount = C;
+                    arg.num_samples = numSamplesInBin;
+                    arg.scale = static_cast<const float*>(&numSamplesInBinInvert);
+                    std::vector<float> workingBuf(rnd_up(C, 16), 0.f);
```

Comment:
Cannot we preallocate memory outside of the parallel section?

Reply:
done, thanks Dmitry.